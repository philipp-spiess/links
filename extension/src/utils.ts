export function parseData(raw: string): [string, string][] {
	const links: [string, string][] = [];
	for (const link of raw.split("\n")) {
		const [path, url] = link.split(/\s+/);
		if (!path || !url || path.startsWith("#") || path === "/") {
			continue;
		}
		links.push([path, url]);
	}
	return links;
}

// The raw text is a list of lines, where each line has two whitespace-separated
// values.
//
// This script ensures that the indentation of the second value is the same
// across all lines by setting the indentation based on the longest first value.
export function syncIndentation(text: string): string {
	const lines = text.split("\n");
	const newLines: string[] = [];

	const groups = findGroups(lines);
	for (const group of groups) {
		const indentation =
			group.reduce((acc, line) => {
				const parsed = parseLine(line);
				if (parsed) {
					return Math.max(acc, parsed[0].length);
				}
				return acc;
			}, 0) + 1;

		for (const line of group) {
			const parsed = parseLine(line);

			if (parsed) {
				const [firstValue, secondValue] = parsed;
				newLines.push(
					`${firstValue}${" ".repeat(
						indentation - firstValue.length,
					)}${secondValue}`,
				);
				continue;
			}
			newLines.push(line);
		}
	}

	return newLines.join("\n");
}

function findGroups(lines: string[]): string[][] {
	let currentGroupHasEntries = false;
	let currentGroup: string[] = [];
	const groups: string[][] = [currentGroup];
	for (const line of lines) {
		const parsed = parseLine(line);
		if (parsed || !currentGroupHasEntries) {
			if (parsed) {
				currentGroupHasEntries = true;
			}
			currentGroup.push(line);
		} else {
			// Start a new group
			currentGroup = [line];
			currentGroupHasEntries = false;
			groups.push(currentGroup);
		}
	}
	return groups;
}

export function isValidLink(link: string): boolean {
	return /\/[a-zA-Z0-9\-_]+/.test(link);
}

export function isValidUrl(url: string): boolean {
	try {
		new URL(url);
		return true;
	} catch {
		return false;
	}
}

function parseLine(line: string): [string, string] | null {
	const [path, url] = line.split(/\s+/);
	if (!path || !url || path.startsWith("#")) {
		return null;
	}
	return [path, url];
}
