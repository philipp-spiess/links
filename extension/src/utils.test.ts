import { describe, it, expect } from "vitest";
import dedent from "dedent";
import { parseData, syncIndentation } from "./utils";

describe("parseDataInSections", () => {
	it("can parse", () => {
		const links = dedent`

			/ https://example.com

			# foo
			/foo https://example.com/foo
			/bar https://example.com/bar

			# baz
			/baz       https://example.com/baz
			/quxquxqux https://example.com/qux
		`;

		expect(parseData(links)).toEqual([
			{
				title: null,
				links: [["/", "https://example.com"]],
			},
			{
				title: "foo",
				links: [
					["/foo", "https://example.com/foo"],
					["/bar", "https://example.com/bar"],
				],
			},
			{
				title: "baz",
				links: [
					["/baz", "https://example.com/baz"],
					["/quxquxqux", "https://example.com/qux"],
				],
			},
		]);
	});
});

describe("syncIndentation", () => {
	it("finds the right indentations per group", () => {
		const text = dedent`
			# 1
			/aaaa https://example.com
			/aaaaaaaa https://example.com
			# 2
			/aaaaaaaaaaaa https://example.com
			/aa https://example.com

			#3
			/aaa https://example.com
			/aaaaa https://example.com
		`;
		expect(syncIndentation(text)).toBe(dedent`
			# 1
			/aaaa     https://example.com
			/aaaaaaaa https://example.com
			# 2
			/aaaaaaaaaaaa https://example.com
			/aa           https://example.com

			#3
			/aaa   https://example.com
			/aaaaa https://example.com
		`);
	});
});
