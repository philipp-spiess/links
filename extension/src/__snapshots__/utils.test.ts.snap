// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseData > can parse 1`] = `
[
  [
    "/foo",
    "https://example.com/foo",
  ],
  [
    "/bar",
    "https://example.com/bar",
  ],
  [
    "/baz",
    "https://example.com/baz",
  ],
  [
    "/quxquxqux",
    "https://example.com/qux",
  ],
]
`;
